### ISSUE TITLE
Wine Fullscreen Apps are not Unredirected Correctly

### ISSUE THREAD

**Issue by li20034** [linuxmint/muffin#701]
### Distribution

Mint 22

### Package version

6.2.9

### Graphics hardware in use

Intel HD 530

### Frequency

Always

### Bug description

Fullscreen apps running in Wine are still being composited likely negatively impacting performance and latency, rather than being unredirected so the app draws directly to the screen.

### Steps to reproduce

1. Turn on `Settings -> General -> Compositor Options -> Disable Compositing for full-screen Windows`
2. Run a 3D demo or game in Wine in fullscreen. Something which runs at a high framerate and vsync off is desirable to really see the effects.
3. ssh into the machine from another device
4. in the ssh session, run `sudo intel_gpu_top`. You will observe that cinnamon is listed with a significant amount of 3D Render usage. Also there is no tearing even with vsync off in the app. This means that cinnamon is still compositing the window.
5. For reference, try running the native app `glmark2 --fullscreen` instead of Wine. This time, the GPU render usage for cinnamon will go to (almost) 0 after a second or 2.

### Expected behavior

Cinnamon should use almost 0 GPU in this scenario, and the window gets unredirected, meaning that with VSync off, tearing should be visible. 

### Additional information

I'm using Wine 9.0 from the official Ubuntu repos (not the winehq ppa). I listed the GPU as Intel HD 530 (on i5 6600), but the same issue also occurs on other Intel GPUs (e.g. Intel Iris Xe on i7-1180G7). I'm almost sure that this issue occurs with other GPUs too, but I do not have the hardware to test it.

**Comment by li20034** [comment]
That's definitely interesting. So it's likely not an issue with Wine/Proton itself then, since KDE works. One thing I'm curious about is whether this happens on Gnome, since cinnamon and muffin are forks of gnome-shell and mutter respectively. I haven't had the chance to test that yet since I don't use Gnome.

The other thing I'm curious about is whether this issue occurred in Linux Mint 21. Because a few years ago, there was a very similar issue to this (in terms of symptoms) described [here](https://github.com/linuxmint/muffin/issues/544), which was fixed in late 2021.

**Comment by Enokilis** [comment]
> The other thing I'm curious about is whether this issue occurred in Linux Mint 21. Because a few years ago, there was a very similar issue to this (in terms of symptoms) described [here](https://github.com/linuxmint/muffin/issues/544), which was fixed in late 2021.

I believe Muffin was recently reforked from upstream Mutter, nullifying most of the code issues from years ago.
I was going to test GNOME too, but it seems to have a regression on X11 that prevents proper page flipping even in native apps. I'd have to try it on Debian stable.

**Comment by Enokilis** [comment]
I can now confirm that GNOME 43 behaves the exact same way as Cinnamon with the fullscreen compositor bypass bool checked, so it doesn't appear to be a Muffin issue specifically.
I doubt the GNOME team has any interest in dealing with X11 issues, though.

**Comment by itz-me-zappex** [comment]
I figured out that composition unredirection is not really broken, something just fails to request it for some windows because of bad logic (for example, unredirection does not work for Minecraft Java Edition despite it runs natively and not through Wine/Proton), you all can try this patch, which enforces unredirection for all fullscreen windows even if that behaviour disabled in settings:
```diff
diff --git a/src/compositor/meta-surface-actor-x11.c b/src/compositor/meta-surface-actor-x11.c
index 7b2a742f..1a11b412 100644
--- a/src/compositor/meta-surface-actor-x11.c
+++ b/src/compositor/meta-surface-actor-x11.c
@@ -265,8 +265,8 @@ meta_surface_actor_x11_should_unredirect (MetaSurfaceActorX11 *self)
   if (window->shape_region != NULL)
     return FALSE;
 
-  if (!meta_window_is_monitor_sized (window))
-    return FALSE;
+  if (meta_window_is_monitor_sized (window))
+    return TRUE;
 
   if (meta_window_requested_bypass_compositor (window))
     return TRUE;
```
And it does work, I can see tearing if vsync disabled in game, but that is just to figure out why unredirection does not work, not a real fix because of reason I described above.

**Comment by itz-me-zappex** [comment]
So, there is multiple functions in different modules which are used to define whether window supposed to be unredireted or not, I found 3 (three) of them (those are seeming to communicate as 3rd -> 2nd -> 1st):
1. `src/x11/window-props.c` -> `reload_bypass_compositor()` -> [1846th line](<https://github.com/linuxmint/muffin/blob/ad96ab1d06251b5f9299317e43ccf28fabad9e76/src/x11/window-props.c#L1846>)
2. `src/core/window.c` ->`meta_window_requested_bypass_compositor()` -> [3170th line](<https://github.com/linuxmint/muffin/blob/ad96ab1d06251b5f9299317e43ccf28fabad9e76/src/core/window.c#L3170>)
3. `src/compositor/meta-surface-x11.c` -> `meta_window_requested_bypass_compositor()` -> [271st line](<https://github.com/linuxmint/muffin/blob/ad96ab1d06251b5f9299317e43ccf28fabad9e76/src/compositor/meta-surface-actor-x11.c#L271>)

And `reload_bypass_compositor()` from `src/x11/window-props.c` seems to rely on `NET_WM_BYPASS_COMPOSITOR` atom (`src/x11/window-props.c` -> `meta_x11_display_init_window_prop_hooks()` -> [2005th line](<https://github.com/linuxmint/muffin/blob/ad96ab1d06251b5f9299317e43ccf28fabad9e76/src/x11/window-props.c#L2005>)), which does not exist on any window when I try get window atoms list through `xprop`.

**Comment by itz-me-zappex** [comment]
TL;DR: `meta_window_requested_bypass_compositor()` does not seem to do anything, at all.

So, there is multiple functions in different modules which are used to define whether window supposed to be unredireted or not, I found 3 (three) of them (those are seeming to communicate as 3rd -> 2nd -> 1st):
1. `src/x11/window-props.c` -> `reload_bypass_compositor()` -> [1846th line](<https://github.com/linuxmint/muffin/blob/ad96ab1d06251b5f9299317e43ccf28fabad9e76/src/x11/window-props.c#L1846>)
2. `src/core/window.c` ->`meta_window_requested_bypass_compositor()` -> [3170th line](<https://github.com/linuxmint/muffin/blob/ad96ab1d06251b5f9299317e43ccf28fabad9e76/src/core/window.c#L3170>)
3. `src/compositor/meta-surface-x11.c` -> `meta_window_requested_bypass_compositor()` -> [271st line](<https://github.com/linuxmint/muffin/blob/ad96ab1d06251b5f9299317e43ccf28fabad9e76/src/compositor/meta-surface-actor-x11.c#L271>)

And `reload_bypass_compositor()` from `src/x11/window-props.c` seems to rely on `NET_WM_BYPASS_COMPOSITOR` atom (`src/x11/window-props.c` -> `meta_x11_display_init_window_prop_hooks()` -> [2005th line](<https://github.com/linuxmint/muffin/blob/ad96ab1d06251b5f9299317e43ccf28fabad9e76/src/x11/window-props.c#L2005>)), which does not exist on any window when I try get window atoms list through `xprop`.

**Comment by itz-me-zappex** [comment]
This patch (for testing purposes only, continue read) enforces unredirection even if that behavior disabled in settings for all fullscreen windows (including fullscreen games running through Wine/Proton):
```diff
diff --git a/src/compositor/meta-surface-actor-x11.c b/src/compositor/meta-surface-actor-x11.c
index 7b2a742f..9d63ae4d 100644
--- a/src/compositor/meta-surface-actor-x11.c
+++ b/src/compositor/meta-surface-actor-x11.c
@@ -268,6 +268,9 @@ meta_surface_actor_x11_should_unredirect (MetaSurfaceActorX11 *self)
   if (!meta_window_is_monitor_sized (window))
     return FALSE;
 
+  if (meta_window_is_monitor_sized (window))
+    return TRUE;
+
   if (meta_window_requested_bypass_compositor (window))
     return TRUE;

```

Using `meta_prefs_get_unredirect_fullscreen_windows()` in added `if` statement (`if (meta_window_is_monitor_sized (window) && meta_prefs_get_unredirect_fullscreen_windows ())`) does not seem to return `true`, at all, even if checking it for being false (`!meta_prefs_get_unredirect_fullscreen_windows ()`). Also it seems to unredirect `nemo-desktop` on my laptop with UHD 600 graphics with that patch, but does not on NVIDIA 1050Ti desktop. Even more, running native games outside of Steam/Lutris (e.g. Minecraft) makes unredirection work flawlessly without that testing patch, also in NVIDIA Settings I can enable API indicator which shows method through which frames are thrown to screen, in Minecraft is `FLIP` as supposed to be, in Proton/Wine games (independently by whether that is OpenGL or Vulkan) - `BLIT` (but now with tearing), and that sucks, because that is still redirection which breaks GSYNC on NVIDIA and causes wild frameskip/slowdown on UHD 600 if FPS higher than screen refresh rate under heavy load (does not happen with Openbox WM w/o compositor), I can not find anything related to that in source code, and I can not get why `meta_prefs_get_unredirect_fullscreen_windows()` ignores specified setting related to unredirection (and I have tried restart Cinnamon using `Ctrl`+`Alt`+`Esc` and relogin).

Tested on Arch Linux, Muffin 6.4.1 built through `makepkg` after modifying source code ([makepkg: Build from local source files](<https://wiki.archlinux.org/title/Makepkg#Build_from_local_source_files>)).

**Comment by itz-me-zappex** [comment]
Issue is that WM unable to obtain size of window created by app which runs through Wine/Proton, so those windows always have `x` and `y` set to `0` and `width` and `height` to `1`, even more unredirect is enabled by default even if that behavior disabled in settings.

Proof that Wine/Proton apps have `x`/`y` set to `0` (allows BLIT unredirect for those apps, even if those are windowed):
```diff
diff --git a/src/core/window.c b/src/core/window.c
index 249e8e3b..6de31ccc 100644
--- a/src/core/window.c
+++ b/src/core/window.c
@@ -3106,6 +3106,13 @@ meta_window_is_screen_sized (MetaWindow *window)
   meta_display_get_size (window->display, &screen_width, &screen_height);
   meta_window_get_frame_rect (window, &window_rect);
 
+  /* Assuming that is a bug which prevents obtaining window size properly of software running through Wine/Proton 
+   * Enforces unredirect for all kind of software which runs through Wine/Proton
+   */
+  if (window_rect.x == 0 && window_rect.y == 0 &&
+      window_rect.width == 1 && window_rect.height == 1)
+    return TRUE;
+
   if (window_rect.x == 0 && window_rect.y == 0 &&
       window_rect.width == screen_width && window_rect.height == screen_height)
     return TRUE;
```

Disallow unredirect if disabled in settings (or through dconf):
```diff
diff --git a/src/compositor/meta-surface-actor-x11.c b/src/compositor/meta-surface-actor-x11.c
index 7b2a742f..0e442010 100644
--- a/src/compositor/meta-surface-actor-x11.c
+++ b/src/compositor/meta-surface-actor-x11.c
@@ -254,6 +254,9 @@ meta_surface_actor_x11_is_opaque (MetaSurfaceActor *actor)
 gboolean
 meta_surface_actor_x11_should_unredirect (MetaSurfaceActorX11 *self)
 {
+  if (!meta_prefs_get_unredirect_fullscreen_windows ())
+    return FALSE;
+
   MetaWindow *window = self->window;
 
   if (meta_window_requested_dont_bypass_compositor (window))
@@ -277,7 +280,7 @@ meta_surface_actor_x11_should_unredirect (MetaSurfaceActorX11 *self)
   if (meta_window_is_override_redirect (window))
     return TRUE;
 
-  if (self->does_full_damage && meta_prefs_get_unredirect_fullscreen_windows ())
+  if (self->does_full_damage)
     return TRUE;
 
   return FALSE;
```

Tested on Arch Linux, NVIDIA 1050Ti (`570.86.16`, proprietary), Muffin v6.4.1.

**Comment by itz-me-zappex** [comment]
I found a more elegant way to do exactly the same, Wine/Proton apps have `window->override_redirect` set to `true`, and it also works as expected (except it also causes unredirection of windowed apps as a patch which checks for `x`/`y`->`0` and `height`/`width`->`1`). (Fixes broken compositor unredirection setting and fails only if window is neither monitor sized nor has override_redirect).
```diff
diff --git a/src/compositor/meta-surface-actor-x11.c b/src/compositor/meta-surface-actor-x11.c
index 7b2a742f..7691302b 100644
--- a/src/compositor/meta-surface-actor-x11.c
+++ b/src/compositor/meta-surface-actor-x11.c
@@ -254,6 +254,9 @@ meta_surface_actor_x11_is_opaque (MetaSurfaceActor *actor)
 gboolean
 meta_surface_actor_x11_should_unredirect (MetaSurfaceActorX11 *self)
 {
+  if (!meta_prefs_get_unredirect_fullscreen_windows ())
+    return FALSE;
+
   MetaWindow *window = self->window;
 
   if (meta_window_requested_dont_bypass_compositor (window))
@@ -265,7 +268,7 @@ meta_surface_actor_x11_should_unredirect (MetaSurfaceActorX11 *self)
   if (window->shape_region != NULL)
     return FALSE;
 
-  if (!meta_window_is_monitor_sized (window))
+  if (!meta_window_is_monitor_sized (window) && !window->override_redirect)
     return FALSE;
 
   if (meta_window_requested_bypass_compositor (window))
@@ -277,7 +280,7 @@ meta_surface_actor_x11_should_unredirect (MetaSurfaceActorX11 *self)
   if (meta_window_is_override_redirect (window))
     return TRUE;
 
-  if (self->does_full_damage && meta_prefs_get_unredirect_fullscreen_windows ())
+  if (self->does_full_damage)
     return TRUE;
 
   return FALSE;
```

Also I should add that I tried first and this patches on UHD 600 graphics laptop, and compositor shuts down completely instead of unredirect window (and compositing is not restored until I close game which runs through Proton). Weird that is does not happen for me on NVIDIA, but here we go, and I do not have AMD GPU to test behavior there.

**Comment by itz-me-zappex** [comment]
Tested it on laptop with UHD 600 graphics, and there is few issues with unredirection of Wine/Proton stuff here:
1. It breaks in-game VSYNC completely.
2. It breaks VSYNC in compositor completely, until I close unredirected window (i.e. exit from game).

Weird, that does not seem to happen on NVIDIA at all, both in-game VSYNC and compositor works fine there, but I should mention that `i915` driver is kinda buggy and I have broken VSYNC even on desktop if I use kernel newer than 6.6 (LTS one).

**Comment by itz-me-zappex** [comment]
It seems I found root of issue, stuff which runs through Wine/Proton has parent window and child one, last has `1x1+0+0` dimensions and rectangles (that does not happen with native games), `xwininfo -tree -root`:
```
xwininfo: Window id: 0x4c6 (the root window) (has no name)

  Root window id: 0x4c6 (the root window) (has no name)
  Parent window id: 0x0 (none)
     119 children:
     ...
     0x6800003 "Geometry Dash": ("steam_app_322170" "steam_app_322170")  1366x912+0+0  +0+0
        1 child:
        0x6600031 (has no name): ()  1x1+0+0  +0+0
    ...
```

Now time to figure what to do with that, no idea tbh.

**Comment by itz-me-zappex** [comment]
Turns out that issue reproducible even with native games launched through Steam, as those also have child window with broken dimensions.

Also it seems that BLIT is prefered over FLIP for such windows (unredirection with BLIT causes issues with Intel graphics I described above and breaks G-Sync/FreeSync on Nvidia and probably AMD), because parent window still appears redirected (while child one is not), that explains why Minecraft Java Edition and glxgears (with -fullscreen) are using FLIP method while unredirected.

I guess that WM should check whether parent window has single child with broken dimensions and unredirect both of them in this case, but idk how to implement that as I am not a developer of muffin and source code is pretty complicated because of project size.

P.S.: Using pure openbox (w/o compositor) in case with UHD 600 works like a charm, i.e. it does not break in-game vsync and does not cause frameskips/slowdowns, probably due to FLIP.

**Comment by itz-me-zappex** [comment]
I figured out that muffin gets and remembers window info immediately when it appears and does not update it ever, and it seems that Wine/Proton/Steam creates "technical" window when game starts, and Cinnamon associates it with main game window causing inablity to obtain its proper size to allow unredirection.

To make sure in that, you (all) just need to start game through Steam (does not matter native or through Proton), then minimize game and run in terminal `xwininfo -tree -root`, there you (all) can find main game window and its child with `1x1+0+0` rects, after that restart Cinnamon (using `Ctrl`+`Alt`+`Escape` for example) and run `xwininfo -tree -root` again. Magically, previously broken child window now has proper size and rectangles, but unredirection still does not work, because Cinnamon uses somewhat "cached" info, which is still `1x1+0+0`, and just ignores unredirection.

This patch fixes that by getting info about focused from scratch, works as expected, unredirection happens only if window has screen size and `x=0`+`y=0`, but that is still does not enforce FLIP method like in case with purely native games (I mean those launched outside of Steam), but at least allows tearing and fixes stuttering caused by compositor.
```diff
diff --git a/src/core/window.c b/src/core/window.c
index 249e8e3b..fc405b1f 100644
--- a/src/core/window.c
+++ b/src/core/window.c
@@ -3106,6 +3106,54 @@ meta_window_is_screen_sized (MetaWindow *window)
   meta_display_get_size (window->display, &screen_width, &screen_height);
   meta_window_get_frame_rect (window, &window_rect);
 
+  /* Workaround, obtain focused window size and rectangles manually if window has 'x' and 'y' set to '0' and size to '1x1'
+   * Fixes issue with inability to unredirect software which runs through Wine/Proton, and even native one which runs through Steam
+   * But that is not perfect solution as still BLIT instead of FLIP being used
+   * Using already obtained info by WM instead of obtaining new one causes the same buggy result
+   */
+  if (window_rect.x == 0 && window_rect.y == 0 &&
+      window_rect.width == 1 && window_rect.height == 1)
+    {
+      Window active_window;
+      Atom actual_type;
+      int actual_format;
+      unsigned long nitems, bytes_after;
+      unsigned char *prop = NULL;
+
+      Display *display = XOpenDisplay(NULL);
+      Window root = DefaultRootWindow(display);
+
+      Atom atom = XInternAtom(display, "_NET_ACTIVE_WINDOW", False);
+
+      if (XGetWindowProperty(display, root, atom, 0, 1, False, XA_WINDOW,
+          &actual_type, &actual_format, &nitems, &bytes_after, &prop) == Success &&
+          prop != NULL)
+        {
+          active_window = *((Window *)prop);
+          XFree(prop);
+
+          if (active_window != None)
+            {
+              Window root, parent;
+              Window *children;
+              unsigned int nchildren;
+              XQueryTree(display, active_window, &root, &parent, &children, &nchildren);
+
+              XWindowAttributes window_attributes;
+              XGetWindowAttributes(display, active_window, &window_attributes);
+
+              window_rect.x = window_attributes.x;
+              window_rect.y = window_attributes.y;
+              window_rect.width = window_attributes.width;
+              window_rect.height = window_attributes.height;
+            }
+        }
+
+      XCloseDisplay(display);
+    }
+  /* End of workaround
+   */
+
   if (window_rect.x == 0 && window_rect.y == 0 &&
       window_rect.width == screen_width && window_rect.height == screen_height)
     return TRUE;
```

**Comment by itz-me-zappex** [comment]
I forgot to add that pull request I have created and patch above does not cause vsync breakage on desktop itself and in any other app _on UHD 600_ graphics when window unredirected, _unlike previous patch which allows unredirection for all windows with `1x1+0+0` size&rects_ (anyway it has been written for testing purposes), but still breaks in-game vsync in unredirected games, _again, only on UHD 600_, works like a charm on NVIDIA. In any case I guess that is better to make unredirection work at least in that way (also pull request fixes ignoreance of setting responsible for redirection, so if someone has issue with that, can safely disable unredirection, and that feature already disabled by dafault), instead of keep it in that state, especially including a fact that nobody from development team works on that.

**Comment by itz-me-zappex** [comment]
Improved patch, I will open pull request again.

```diff
diff --git a/src/core/window.c b/src/core/window.c
index 249e8e3..b8ae9e1 100644
--- a/src/core/window.c
+++ b/src/core/window.c
@@ -3106,6 +3106,65 @@ meta_window_is_screen_sized (MetaWindow *window)
   meta_display_get_size (window->display, &screen_width, &screen_height);
   meta_window_get_frame_rect (window, &window_rect);
 
+  /* Workaround, obtain currently focused window rectangles if window has 'x' and 'y' set to '0' and size to '1x1' (i.e. child with broken rectangles)
+   * Fixes issue with inability to unredirect fullscreen games running through Wine/Proton, and even native ones running through Steam
+   * In such cases parent window accepts input while child one responsible for output, that is why parent appears skipped (and still redirected) but child is handled
+   * To make sure in that, run any game through Steam, run 'xwininfo -tree -root', find game window there and unmap child window using 'xdotool windowunmap <child>'
+   * After that switch back to game (screen becomes blank) and press any random buttons, then map window back ('xdotool windowmap <child>') and you will see that game handled input
+   * That is not perfect solution as still BLIT instead of FLIP being used (due to reason described above), but at least that allows tearing and removes hiccups caused by compositor
+   * Unredirection with BLIT is known to break in-game VSYNC completely on Intel Graphics
+   * But it works fine in case with NVIDIA (at least with proprietary driver) and do not have AMD GPU to test
+   * Using ID of focused window (i.e. parent) is needed to avoid the same buggy rectangles
+   */
+  if (window_rect.x == 0 && window_rect.y == 0 &&
+      window_rect.width == 1 && window_rect.height == 1)
+    {
+      Window active_window;
+      Atom actual_type;
+      int actual_format;
+      unsigned long nitems, bytes_after;
+      unsigned char *prop = NULL;
+
+      MetaX11Display *x11_display = window->display->x11_display;
+      Window root = meta_x11_display_get_xroot (x11_display);
+      Display *xdisplay = meta_x11_display_get_xdisplay (x11_display);
+
+      Atom net_active_window = XInternAtom (xdisplay, "_NET_ACTIVE_WINDOW", False);
+
+      if (XGetWindowProperty (xdisplay, root, net_active_window, 0, 1, False, XA_WINDOW,
+          &actual_type, &actual_format, &nitems, &bytes_after, &prop) == Success &&
+          prop != NULL &&
+          actual_type == XA_WINDOW &&
+          nitems == 1)
+        {
+          active_window = *((Window *)prop);
+
+          XFree (prop);
+
+          if (active_window != None)
+            {
+              Window parent;
+              Window *children = NULL;
+              unsigned int nchildren;
+              XQueryTree (xdisplay, active_window, &root, &parent, &children, &nchildren);
+
+              if (children)
+                XFree (children);
+
+              XWindowAttributes window_attributes;
+              if (XGetWindowAttributes (xdisplay, active_window, &window_attributes))
+                {
+                  window_rect.x = window_attributes.x;
+                  window_rect.y = window_attributes.y;
+                  window_rect.width = window_attributes.width;
+                  window_rect.height = window_attributes.height;
+                }
+            }
+        }
+    }
+  /* End of workaround
+   */
+
   if (window_rect.x == 0 && window_rect.y == 0 &&
       window_rect.width == screen_width && window_rect.height == screen_height)
     return TRUE;
```

**Comment by itz-me-zappex** [comment]
Turns out that unredirection is impossible for fullscreen Wine/Proton/native Steam games __using OpenGL__ (vanilla muffin 6.4.1), patch for `src/core/window.c` allows BLIT unredirection for those OpenGL games, but the most interesting thing is that muffin already unredirects Vulkan apps perfectly (meaning allows FLIP), except unredirection will work even if that disabled in settings (check `src/compositor/meta-surface-actor-x11.c` patch).

And `xwininfo -tree -root` shows that Vulkan games are having proper rectangles on child windows, unlike OpenGL ones with `1x1+0+0`, weird.

So I could assume that something else also behaves weird on muffin side, since if unredirection allowed, it should use FLIP for OpenGL too, but that very hard to figure with that out.

**Comment by itz-me-zappex** [comment]
Finally, I fixed that in proper way, works like a charm for me w/o exceptions.

**Comment by Enokilis** [comment]
Awesomely done. Native Steam games were never an issue for me, but I also have AMD. Are you using the modesetting driver?

**Comment by itz-me-zappex** [comment]
I have PC with Nvidia GPU and laptop with Intel UHD 600 (yeah, `modesetting`, `xf86-video-intel` shitty as f), about Nvidia, `nvidia-settings` has option to enable API indicator which except FPS and API shows method through which frames are thrown to screen.

![Image](https://github.com/user-attachments/assets/db44f891-b1b8-477c-87db-5f93b5e8ae91)

Works with both Vulkan and OpenGL, that is how it looks like:

![Image](https://github.com/user-attachments/assets/9321e24e-0e05-485e-9407-7f93b5b1f25c)

And before patch which ignores (i.e. does not override previous ones with new) rectangles in case window size reported as `1x1`, fullscreen OpenGL games running through Proton and even native ones running from Steam used `BLIT` instead of `FLIP`, but `glxgears -fullscreen` and Minecraft Java Edition used `FLIP` method as supposed to be.

About `meta-surface-actor-x11.c`, I noticed that `FLIP` method was used for fullscreen stuff even if that setting disabled:

![Image](https://github.com/user-attachments/assets/54aca1ef-7bc6-4a00-b642-9323f4c0ac2a)

And I doubt that is how it supposed to work, change I did there just to prevent unredirection if this setting disabled.

**Comment by itz-me-zappex** [comment]
I use patched muffin, does not seem to cause issues for me with that, I can not upload demo here for unknown reason (`<!-- Failed to upload "2025-02-17 13-25-04.mp4" -->`), so I have uploaded it to [Streamable](https://streamable.com/vzj8bu).

Fullscreen VLC uses `FLIP` instead of `BLIT` but with `VSYNC ON`, so no issue with that either.

Could you build muffin with those patches and check how does it work for you?
